### ISSUE TITLE
A second version!

### ISSUE THREAD

**Issue by rreusser** [scijs/ndarray-lup-factorization#1]
I'm also in the process of implementing this. For reference just in case you're checking github and not email, I have:

```javascript
'use strict'

var blas = require('ndarray-blas-level1')

function lup ( A, L, P ) { 
  var i,j,k, Ljk, Uk, Ukk, m=A.shape[1], U=A, tmp 

  for(i=0;i<m;i++) {
    P[i] = i 
  }

  for(k=0; k<m; k++) {
    L.set(k,k,1)

    i = blas.iamax(U.pick(null,k).lo(k).hi(m-k))+k

    Uk = U.pick(k,null)
    blas.swap( Uk.lo(k).hi(m-k), U.pick(i,null).lo(k).hi(m-k) )
    blas.swap( L.pick(k,null).hi(k), L.pick(i,null).hi(k) )

    tmp=P[i], P[i]=P[k], P[k]=tmp

    Ukk = U.get(k,k)
    for(j=k+1; j<m; j++) {
      Ljk = U.get(j,k)/Ukk
      L.set(j,k,Ljk)
      blas.axpy(-Ljk, Uk.lo(k).hi(m-k), U.pick(j,null).lo(k).hi(m-k))
    }   
  }

  return [ L, U, P ] 
}

module.exports = lup 
```

**Comment by rreusser** [comment]
If you prefer, I also have a version that avoids one of the swaps, but there's obviously swapping required in one place or another, so I'm not entirely sure what's best. I think the L swap can also be avoided, but then you have to do some permutation gymnastics (upcoming module: scijs-inverse-permutation) when solving. In the end, I went with the original version since the result is a totally straightforward L, U, and P with no surprises or weird semi-swapped madness. I think maybe the minimal-swap version would be nice for an LUPSolve module that takes care of the whole thing in one go whereas the swapped version is nice if you just want a friendly answer.

```javascript
function lupWithoutSwap ( A, L, P ) {
  var i,j,k, m=A.shape[1], U=A, tmp
  var Ljk, Uk, Ukk
  for(i=0;i<m;i++) { P[i] = i }
  var umax, imax,tmp;
  for(k=0; k<m; k++) {
    L.set(k,k,1)
    for(imax=k,umax=-Infinity,i=k;i<m;i++) {
      if( (tmp=Math.abs(U.get(P[i],k))) > umax ) {
        umax = tmp, imax=i
      }
    }
    i = imax
    Uk = U.pick(k,null)
    blas.swap( L.pick(k,null).hi(k), L.pick(i,null).hi(k) )
    tmp=P[i], P[i]=P[k], P[k]=tmp
    Ukk = U.get(P[k],k)
    for(j=k+1; j<m; j++) {
      Ljk = U.get(P[j],k)/Ukk
      L.set(j,k,Ljk)
      blas.axpy(-Ljk, U.pick(P[k],null).lo(k).hi(m-k), U.pick(P[j],null).lo(k).hi(m-k))
    }
  }

  return [ L, U, P ]
}
```

**Comment by Planeshifter** [comment]
Nice! Here is what I would have had to offer so far: 

``` javascript
function lup( A, pi ) {
	var i, j, k, kp, l, n, p, max, temp, akk;

	n = A.shape[0];

	for ( i = 0; i < n; i++ ) {
		pi[i] = i;
	}

	for (k = 0; k < n; k++) {

		kp = blas.iamax( A.pick( null, k ).lo( k ).hi( n - k ) ) + k;

		if ( kp !== k ) {
			Ak = A.pick( k, null );
			blas.swap( Ak.lo( k ).hi( n - k ), A.pick( kp, null ).lo( k ).hi( n - k) );
			temp = pi[k];
			pi[k] = pi[kp];
			pi[kp] = temp;
		}
		akk = A.get( k, k );
		if (akk !== 0.0) {
			for ( i = k + 1; i < n; i++ ) {
				A.set( i, k, A.get( i, k ) / akk );
				for ( j = k + 1; j < n; j++ ) {
				  A.set( i, j, A.get( i, j ) - A.get( i, k ) * A.get( k, j ) );
				}
			}
		}
	}
	return A;
}
``` 
It calculates the LU in-place at the upper- and lower triangular matrices of `A` -  I adapted the algorithm described in the standard textbook "Introduction to Algorithms". Your implementation is better, so feel free to take over the repo.

**Comment by rreusser** [comment]
I mean you have to admit the timing is pretty hilarious… Hmm… I have mixed feelings. Our implementations look pretty close to similar except I think I have one more swap. I'm using Trefethen and Bau's Numerical Linear Algebra and my result ends up just fine, but up next is to use two steps of [ndarray-blas-trsv](https://github.com/scijs/ndarray-blas-trsv) and a permutation to actually solve/invert. If the intermediate result isn't important, I think there might be a shortcut in which you can skip actually swapping and just use it correctly later on. Mixed feelings. I'll give solving a shot and see how this turns out.

**Comment by Planeshifter** [comment]
Yeah, quite a coincidence. Should have coordinated better, with your amount of output I should have figured that you might be already working on it. :)

**Comment by rreusser** [comment]
Haha, yeah. I'm kinda throwing stuff at the wall to see what sticks, so I really appreciate your effort and everyone else's input too! I had a bit of a lull the last couple days… started playing with the webaudio api to resurrect some realtime tempo detection thing I was working on a while ago. Thought maybe it'd maybe make a good demo, plus I had an attack of existential angst about spending this much time working on linear algebra whatnot without some connection to any real goal or application. So thought maybe it'd be good to give scijs a spin and see what it's like to actually use it. The biggest uncertainty to me at the moment: compute.io and scijs have a lot of overlap… it almost feels like scijs should just retain ndarrays + ndarray algorithms and compute.io should absorb the rest of the effort. But that's definitely out of the scope of this ticket :smile:
